external/PackedCSparse/PackedChol.h

```diff
@@ -0,0 +1,329 @@
+// VolEsti (volume computation and sampling library)
+
+// Copyright (c) 2012-2018 Vissarion Fisikopoulos
+// Copyright (c) 2018 Apostolos Chalkis
+// Copyright (c) 2022 Ioannis Iakovidis
+
+// This file is converted from PolytopeSamplerMatlab 
+//(https://github.com/ConstrainedSampler/PolytopeSamplerMatlab/blob/master/code/solver/PackedCSparse/PackedChol.h) by Ioannis Iakovidis
+
+#pragma once
+#include "Eigen/Eigen"
+#include "SparseMatrix.h"
+#include "chol.h"
+#include "leverage.h"
+#include "leverageJL.h"
+#include "multiply.h"
+#include "qd/dd_real.h"
+#include <random>
+#include <vector>
+using namespace PackedCSparse;
+
+template <typename Tin, typename Tout>
+void get_slice(Tout *out, Tin *in, size_t n, size_t idx) {
+  for (size_t j = 0; j < n; j++)
+    out[j] = double(get(in[j], idx));
+}
+
+template <typename Tin, typename Tout>
+void set_slice(Tout *out, Tin *in, size_t n, size_t idx) {
+  for (size_t j = 0; j < n; j++)
+    set(out[j], idx, double(in[j]));
+}
+
+template <int k, typename Ti> struct PackedChol {
+  using Tx = double;
+  using Tx2 = FloatArray<double, k>;
+  using Te = dd_real;
+
+  // parameters
+  SparseMatrix<Tx, Ti> A;
+  SparseMatrix<Tx, Ti> At;
+  UniqueAlignedPtr<Tx2> w;
+  Tx accuracyThreshold = 1e-6;
+  std::vector<size_t>
+      exactIdx; // k size array. Indices we perform high precision calculation
+  std::vector<size_t>
+      numExact; // number of times we perform high precision decompose (length
+                // k+1, the last one records how many times we do decompose)
+  bool decomposed = false;
+
+  // preprocess info for different CSparse operations (PackedDouble)
+  MultiplyOutput<Tx2, Ti> H;         // cache for H = A W A'
+  CholOutput<Tx2, Ti> L;             // cache for L = chol(H)
+  LeverageOutput<Tx2, Ti> diagP;     // cache for L = chol(H)
+  LeverageJLOutput<Tx2, Ti> diagPJL; // cache for L = chol(H)
+
+  // preprocess info for different CSparse operations (dd_real)
+  MultiplyOutput<Te, Ti> H_exact;         // cache for H = A W A'
+  CholOutput<Te, Ti> L_exact;             // cache for L = chol(H)
+  LeverageOutput<Te, Ti> diagP_exact;     // cache for L = chol(H)
+  LeverageJLOutput<Te, Ti> diagPJL_exact; // cache for L = chol(H)
+  SparseMatrix<Te, Ti> Le[k];             // store output of L_exact
+  typedef Eigen::Matrix<Tx, Eigen::Dynamic, Eigen::Dynamic> MT;
```

Comment:
a mixing of `using` and `typedef`, prefer `using` as more modern

Reply:
Fixed it, it slipped.